### ISSUE TITLE
Lanes missing in OSI output of OSMP FMU

### ISSUE THREAD

**Issue by ClemensLinnhoff** [esmini/esmini#639]
Lanes are completely missing from the OSI ground truth in the OSMP FMU.
But both lanes and lane boundaries used to be there in previous releases of esmini.

I will further investigate.

**Comment by eknabevcc** [comment]
Do you happen to have a version nr, the more recent the better, of a previous release where it worked? 

If the changed behavior happened recently we'll definitely investigate. Otherwise it may be related to the optimization you mention. The idea there is to communicate static data only once, upfront, while dynamic data is added and updated continuously.

Do you have a use case or scenario example where the optimization scheme makes no sense?

**Comment by ClemensLinnhoff** [comment]
I had a very old trace file lying around that was generated with esmini, where the lanes were present. So unfortunately I don't know the version number. But it was some time ago.

I still have to figure out, why the lanes are not present in the first time step. Because even with the optimization they are not there. I commented out the `if (GetCounter() == 0` and the `ClearOSIGroundTruth()` and then I have the lanes in every time step. So generally it should work.

What I am unsure about is, if this is standard compliant. I know that we discussed this generally in OSI, but I don't know if it made it into the specification. I could not find it in the OSI documentation but that does not mean that it really is not in there ;).
Do you know if and where this is described in the OSI specification?

I don't have an immediate use-case. But we are implementing lanes and lane_boundaries in our visualization tool Simspector and I wanted to use esmini to generate some test data.

So the behavior, that the lanes are completely missing from the SensorView outputted by the OSMP FMU we definitely need to fix. The optimization, we probably need to think about a bit more.

**Comment by eknabevcc** [comment]
Can't reproduce. 
Make sure to call SE_Close() between each run/simulation.
If that doesn't help, we need more details in order to further investigate.

**Comment by ClemensLinnhoff** [comment]
I found out, why the lanes are missing in the first time step in the FMU. This came from calling SE_UpdateOSIGroundTruth(); twice. Once during initialization and then again in the first actual time step.
I fixed that in #643.

But the question remains, whether it is OSI-compliant to send the stationary data only once in the beginning.
For the GroundTruth I am not entirely sure. But for the SensorView I would say it is not compliant, because a SensorView may only contain partial GroundTruth data, so stationary object may be added and removed during simulation.

@jdsika what do you think?

**Comment by ClemensLinnhoff** [comment]
@eknabevcc what would be the best way to add the possibility to the API to configure whether to send static data in all time steps or only the first one?

**Comment by eknabevcc** [comment]
I agree an option is a good compromise solution in this case!

Something like: `--always_included_static_osi`

This way the default behavior will remain, reducing overhead of sending duplicate data for (probably) the majority of cases.

BTW: Options can be set (and unset) via the API as well, see [User guide - Program Options](https://esmini.github.io/#_program_options).

Would this be OK?

**Comment by jdsika** [comment]
I will note that the "OSI standard" does include all data and therefore I would do it exactly the other way around with the option OR at least give a note at startup like "this simulator reduces data by default, for standard conform behavior use option xy"

Otherwise you might teach them the wrong behavior as default?

**Comment by eknabevcc** [comment]
I agree regarding sensorView. We'll look into it, how it could be supported in a no-penalty way to candidate for default behavior. 

BTW: Is there a mechanism in OSI to signal data updates, e.g. like the [dirty bit](https://en.wikipedia.org/wiki/Dirty_bit) concept?

**Comment by ClemensLinnhoff** [comment]
> BTW: Is there a mechanism in OSI to signal data updates, e.g. like the [dirty bit](https://en.wikipedia.org/wiki/Dirty_bit) concept?

Not that I am aware of.

**Comment by eknabevcc** [comment]
In esmini [v2.45.0](https://github.com/esmini/esmini/releases/tag/v2.45.0) it is now possible to request full groundtruth, including static data, at any time.

See release notes (and further info linked from there).

Hopefully it covers immediate need.

**Comment by thomassedlmayer** [comment]
OSI offers the `omit_static_information` flag in [SensorViewConfiguration](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/latest/gen/structosi3_1_1SensorViewConfiguration.html#a124ebfbab7292e8ab9b99fd70d41d797) to indicate (from simulation environment point of view) or request (from sensor model point of view) that the simulation environment provides or omits static information. But this then requires a "GroundTruthInit" message containing all static data at simulation initialization. It is not defined by OSI what is considered static or dynamic, so that's to decide by the simulator. Maybe this is helpful?

**Comment by ClemensLinnhoff** [comment]
> In esmini [v2.45.0](https://github.com/esmini/esmini/releases/tag/v2.45.0) it is now possible to request full groundtruth, including static data, at any time.

In the `SE_UpdateOSIGroundTruth()` function call I set refetchStaticGt to true (see #668). But unfortunately, this did not change anything. The lanes and also the model_reference of the stationary environment are only present in the first time step.

I attached a human-readable OSI trace file that I generated with the esmini FMU from #668.

[20250205T093254Z_sv_370_2112_140_simple_road_two_vehicles.zip](https://github.com/user-attachments/files/18669049/20250205T093254Z_sv_370_2112_140_simple_road_two_vehicles.zip)

**Comment by eknabevcc** [comment]
@slundel6 Can you have a look?

**Comment by slundel6** [comment]
It is true that UpdateOSIStaticGroundTruth() is only called once, when GetCounter() == 0. What this does is that it populates the internal struct of static ground truth (defined [here](https://github.com/esmini/esmini/blob/e9032d96c879bf50f6dc97f370c96515d17ff822/EnvironmentSimulator/Modules/ScenarioEngine/SourceFiles/OSIReporter.cpp#L77) . The internal struct is then referenced by the external osi struct, defined [here](https://github.com/esmini/esmini/blob/e9032d96c879bf50f6dc97f370c96515d17ff822/EnvironmentSimulator/Modules/ScenarioEngine/SourceFiles/OSIReporter.cpp#L84) in [this](https://github.com/esmini/esmini/blob/e9032d96c879bf50f6dc97f370c96515d17ff822/EnvironmentSimulator/Modules/ScenarioEngine/SourceFiles/OSIReporter.cpp#L282) method, the external struct is then serialized, written to log-file and sent to potential receivers on UDP, then the static data is cleared from the external ground truth struct. 

If `refetchStaticGt == true` and `GetCounter() > 0` then [this](https://github.com/esmini/esmini/blob/e9032d96c879bf50f6dc97f370c96515d17ff822/EnvironmentSimulator/Modules/ScenarioEngine/SourceFiles/OSIReporter.cpp#L365) line of code is executed which will re-populate the external struct with the data from the internal struct (which remains same as it was on `GetCounter() == 0`. This means that static data is available again via API or by UDP, but not written to logfile. 

Worth noting then is that the static data is only parsed from file once, as we assume that the static data is not changed throughout the simulation.

That is at least the intended behavior, please let us know if it doesn't work as you'd expect.

**Comment by ClemensLinnhoff** [comment]
But when I get the GroundTruth data with `SE_GetOSIGroundTruthRaw()` the static elements are not part of the returned data. Do I need to set another flag somewhere to get that? I just saw the flag for the UpdateOSIStaticGroundTruth() function in the documentation. But setting that did not make a difference in the output.

**Comment by slundel6** [comment]
Interesting. Can you provide any more details how the code looks like on your side when the data is not there? We have tried it on our side (and it is tested in CI as well), and it seems static data is there. [This](https://github.com/esmini/esmini/blob/master/EnvironmentSimulator/code-examples/osi-groundtruth/osi-groundtruth.cpp) is how we have done a reference-implementation, if that is followed it should work, so it would be really interesting to understand what goes wrong in your case.

**Comment by ClemensLinnhoff** [comment]
I have the problem with the esmini OSMP FMU. [Here](https://github.com/Persival-GmbH/esmini/blob/658-osmp-fmu-broken-due-to-API-change/OSMP_FMU/esmini.cpp#L254) I added the flag to the update and a couple of lines below I get the GroundTruth. A couple of comments above you can find a trace file that was generated with this FMU.

**Comment by slundel6** [comment]
How are you using the static GT in your example? I can see that GT is serialized [here](https://github.com/Persival-GmbH/esmini/blob/7c886d0407414586a86f145b4185be7c57ad6ce1/OSMP_FMU/esmini.cpp#L260), then copied to `currentGT` [here](https://github.com/Persival-GmbH/esmini/blob/7c886d0407414586a86f145b4185be7c57ad6ce1/OSMP_FMU/esmini.cpp#L271) , then `currentGT` is not used anymore.

How do you check the data?

I made a very simplistic example trying to emulate roughly the code flow in the FMU:
`int main(int argc, char* argv[])
{
    (void)argc;
    (void)argv;
    const osi3::GroundTruth* gt;

    SE_EnableOSIFile(0);  // 0 or "" will result in default filename, ground_truth.osi
    SE_Init("../../resources/xosc/cut-in.xosc", 0, 1, 0, 0);
    gt = reinterpret_cast<const osi3::GroundTruth*>(SE_GetOSIGroundTruthRaw());
    for (int i = 0; i < 3; i++)
    {
        SE_StepDT(0.01f);
        SE_UpdateOSIGroundTruth(true);
        gt = reinterpret_cast<const osi3::GroundTruth*>(SE_GetOSIGroundTruthRaw());
        printf("lane boundaries: %d\n", gt->lane_boundary_size());
        for (int j = 0; j < gt->lane_boundary_size(); j++)
        {
            printf("  lane boundary %d, nr of boundary points: %d\n", j, gt->lane_boundary(j).boundary_line_size());

        #if 0  // change to 1 in order to print all boundary points
                            for (int k = 0; k < gt->lane_boundary(j).boundary_line_size(); k++)
                            {
                                    printf("    (%.2f, %.2f)\n",
                                            gt->lane_boundary(j).boundary_line(k).position().x(),
                                            gt->lane_boundary(j).boundary_line(k).position().y());
                            }
        #endif
        }
    }
    SE_Close();
    return 0;
}`
Here I get the data each frame (at least lane boundaries)...

**Comment by slundel6** [comment]
Ah I see, thanks. I'll have a look at your debug code, and I'll also try to reproduce the error on my side.

**Comment by slundel6** [comment]
I think I might have found the issue, a logic error in OSIReporter.cpp. 

I see in your code that you don't call `SE_EnableOSIFile()`, nor is it communicating via UDP.  Both completely valid options of course. When I removed the statement of `SE_EnableOSIFile(0)` in my example I got same behavior as you.

I looked in to the logic of when we re-load the static data in OSIReporter.cpp, and the if statement there is wrong I believe, [line](https://github.com/esmini/esmini/blob/e9032d96c879bf50f6dc97f370c96515d17ff822/EnvironmentSimulator/Modules/ScenarioEngine/SourceFiles/OSIReporter.cpp#L363).

We should of course load the data into the struct no matter if `GetUDPClientStatus() == 0 || IsFileOpen()`... now it will only load the data if any of those are true options are true. The data should of course be loaded into the struct upon request, but only serialized if either `GetUDPClientStatus() == 0` or  `IsFileOpen()` is true.

I re-arranged the code and now that particular case works for me. I made a push to a [branch](https://github.com/esmini/esmini/tree/feature/osi_static_data_fix) (as I type this it's not green in CI).

If you could either checkout the branch and see if it fixes your issue, alternatively, add a line in your code `SE_EnableOSIFile(0)`, somewhere in the init-stage and see if you get the expected behavior.

**Comment by ClemensLinnhoff** [comment]
I merged your branch into mine and that fixes it. Thank you!
Could you let me know, once the fix is merged into dev? Then I can get my branch ready to be merged.

Or alternatively, you can just set the flag to true in the update function of the FMU, like [here](https://github.com/esmini/esmini/pull/668/files#diff-ef71481c5ce7f4915e80aadcc291e41d5de382adb47f5f5306e139e01718032fR254) in your branch. Then we don't need my PR at all.

**Comment by ClemensLinnhoff** [comment]
Or we merge my PR, which now also contains your changes, since I merged your branch into mine: https://github.com/esmini/esmini/pull/668

**Comment by slundel6** [comment]
Great that you managed to test it, thanks!

We did some further reviewing of the current code and found some more optimizations to be done, and a test-case to add to increase robustness. If its all the same to you, would it be OK to go with your first proposal, where I update the flag in the FMU, as I need to iterate a couple of more times on the branch? If so we can skip the small PR you made.

We then have it merged on dev today and have a release tomorrow if all goes according to plan.

And thank you for bringing up the issue and for the quick feedback.

**Comment by ClemensLinnhoff** [comment]
Sure, thanks for integrating it so quickly. I will close the PR and look forward to the new release :).

**Comment by ClemensLinnhoff** [comment]
Once you have done the merge, I will test it again and if everything works, I will close this issue as completed.

**Comment by eknabevcc** [comment]
Included in [v2.45.2](https://github.com/esmini/esmini/releases/tag/v2.45.2)

**Comment by eknabevcc** [comment]
> OSI offers the `omit_static_information` flag in [SensorViewConfiguration](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/latest/gen/structosi3_1_1SensorViewConfiguration.html#a124ebfbab7292e8ab9b99fd70d41d797) to indicate (from simulation environment point of view) or request (from sensor model point of view) that the simulation environment provides or omits static information. But this then requires a "GroundTruthInit" message containing all static data at simulation initialization. It is not defined by OSI what is considered static or dynamic, so that's to decide by the simulator. Maybe this is helpful?

Interesting information. Will check and consider for future.

**Comment by ClemensLinnhoff** [comment]
> Included in [v2.45.2](https://github.com/esmini/esmini/releases/tag/v2.45.2)

Perfect, thanks. I tested the new release and it works :).

**Comment by thomassedlmayer** [comment]
@eknabevcc Will this also be integrated in the OSI GroundTruth export feature of esmini (using `--osi_file parameter)`? As far as I can tell, there is no option yet in v2.45.2 to include the road network in every frame (like `--always_included_static_osi` as you proposed somewhere above). Or did I overlook something? (I just tried the prebuilt binaries for windows.)

By the way, from OSI perspective, I agree with @jdsika that including everything should be the default because otherwise esmini is basically not sending/exporting a valid OSI file. The lanes will just disappear after the first frame if a tool assumes that a valid OSI file is received. This will especially lead to issues when assigned_lane_ids are present but the corresponding lanes are not.

A bit off topic: `--osi_file` also generates a GroundTruth trace with missing version numbers. Don't know if this is the case for other ways to extract OSI from esmini.

**Comment by slundel6** [comment]
I'll take the liberty to answer for now (I'm in same team as @eknabevcc).

You are correct @thomassedlmayer , the static OSI data is only logged once in the file with no option to toggle that and no other options. We could revisit that topic and see if we should make some logic to have such option if there are use-cases / applications which rely on log files having this data every frame.

I'm sure it has been discussed before, but the reason to only include static data once is because the information is redundant, as the static data doesn't change throughout the simulation (at least not currently), so for big road-networks there is a lot of computational and storage overhead with no gain. 

Thanks for the off topic notice, we'll check it out.

**Comment by eknabevcc** [comment]
As long as static data is not changing throughout simulation session (which currently is the case), we see no reason to include it more than once in log file. 
For any further consideration, please clarify use case.

**Comment by thomassedlmayer** [comment]
I mean, the use case is to comply with the ASAM OSI standard when writing trace files labelled as OSI file, right? It should be possible to use these trace files with other tools that consume OSI files without any workarounds. This is not the case right now because you would have to copy the road network to every frame to have a valid OSI trace.

I understand that this creates redundant data but that's how the simulation interface is defined for now. Currently the only option to omit static information is an exchange of SensorViewConfiguration messages between simulation components (+ subsequent agreement) and a corresponding GroundTruthInit message (which clarifies which fields are actually considered static) as I explained above.

But for trace files this currently can not be indicated properly in the file itself. If things considered static are omitted from the files based on non-standard assumptions, they can not be reliably used with other tools. Tools that consume OSI trace files really shouldn't assume by default that "no data" means that the data from the previous frame applies. How to differentiate actual "no data" from that? How should an interpreter of the trace file know what is actually considered static and what not? E.g. if a vehicle is present in the first frame but not in the second frame; should it be considered static like a parked car? This makes it very unclear and you would have to understand exactly what esmini is doing internally.

To fix the issue with too large road networks to be included in every frame, gt-gen-simulator for example uses chunking to only include parts of the road network around the ego vehicle and modify the road network content after some 100 frames (depending on ego vehicle movement).

In my opinion, if no changes are made to fix this, it should be clearly noted that these exported files are "special" OSI files and do not comply with the standard and that workarounds are needed to use these with OSI-compliant tools. Otherwise people using these files as a reference might assume that this is valid.

**Comment by eknabevcc** [comment]
If the standard assumes the same static data to be included in every trace file frame, then there is an immediate need to work on the standard in this part. We can consider join such an initiative.

The fragmentation of road network sounds like a potential good idea, but should also be standardized, I guess.

I think the above features (simple compression and fragmentation) nicely addresses the rising need for larger scenarios, both in terms of geographic size and complexity.

Meanwhile, we’ll not degrade default behavior of esmini just to comply with a shortcoming of a standard. But yes, I agree it could be mentioned, I think it already is (in user guide OSI section). But we will check next week if something can be clarified. 

Finally, if we see a real use case/need, then of course we can consider adding an option to enforce static data each frame.

**Comment by jdsika** [comment]
I will think about an idea to indicate static information to be present in the first frame only in mcap multi trace files and indicate that properly. I think it would boost rendering performance as well. 
I guess the first step is to indicate what information is allowed to be static in a simulation run. I will come back with an idea

**Comment by pmai** [comment]
So I'm not privy to the underlying seemingly longrunning discussion, or all of the details of how esmini tries to deal with OSI or for what purposes. I have no dedicated opinions on that.

I just want to clarify a couple of points here as regards what the standard says:

> If the standard assumes the same static data to be included in every trace file frame, then there is an immediate need to work on the standard in this part. We can consider join such an initiative.

The standard does not "assume" anything here, it specifies what shall and shall not be part of SensorView (or GroundTruth) messages. And the base requirement is to include all relevant data around the ego/sensor position in each frame, as configured via SensorViewConfig or other mechanisms. There are multiple reasons for this, one of them being that for sensor simulations that is what sensor models usually expect and are geared to work against - and especially for larger scenarios is somewhat efficient due to the location-based filtering (see below).

One other reason is that there are varying views on what is considered static information across use cases and implementations, and it is not trivial to come to one common definition that works for everyone on this. Consider as examples the parked car situation on one end of the spectrum, and newly created road works in longer running simulations on the other end (e.g. the vehicle that lays new lane boundaries driving before the ego vehicle). This gets even more complex when considering simulation objects that have some static and some dynamic aspects, as that then leads to per field variability info that would need to be kept track.

All of this adds complexity in both the sender and receiver of information that are needed to reassemble to regain the consistent view that is needed for further processing - which is something that is deemed best to avoid to reduce the potential for bugs in validation use cases.

For other use cases (usually non-sensor ones, like traffic participant models, or visualization interfaces), OSI/OSMP offers additional ways of handling static information, too:

- There is the GroundTruthInit and related omit_static_information SensorViewConfig option which allows the transmission of static data prior to simulation startup, and optionally the omission in the SensorView message stream. This however requires the static data to never be included in the SensorView message stream, i.e. not even in the first message, but rather transmitted as a seperate GroundTruthInit message (one of the reasons being that it is now no longer sensor-specific, the other that this allows all costly initialization to happen prior to runtime, which is important for real-time use cases).

- For visualization and other purposes of transmitting OSI data where incremental updates make the most sense, we have the StreamingUpdate messages, which provide for incremental updates to be sent, which is much more efficient and flexible than the "static vs. dynamic" assumptions.

These cater to different use cases, which are less localized than sensor models, hence would not benefit from location-based filtering, and often have less strict consistency needs.

The above goes for the exchange of OSI at runtime and hence also translates to the messages that would be stored in an OSI trace file, as there is consistency there. The upcoming MCAP-based multi-trace files will help in storing multiple streams in one file, and provide more compression options than just lzma, as well as new meta-data, but do not change the basic picture.

> The fragmentation of road network sounds like a potential good idea, but should also be standardized, I guess.

It is standardized to the extent that for a SensorView only the relevant "cookie" around the ego/sensor position (as per the requested field of view and range in the SensorViewConfig) needs to be transmitted. It is conciously left to the implementations  how to handle this, as that very much depends on the way they are internally handling the data/scene, and is considered a purely optimization issue (and not a correctness issue). So implementations can still provide everything - which can be good enough for smaller scenarios/maps, they can provide a somewhat larger area for which they have simple mechanisms of selection - or can share across multiple sensors, or they can try to more precisely narrow it down.

BTW: Allowing such optimizations and other things for both seemingly static and dynamic data makes it mandatory that data newly absent is not treated as remaining valid.

> Meanwhile, we’ll not degrade default behavior of esmini just to comply with a shortcoming of a standard.

This is not a shortcoming of the standard, in the sense that something was forgotten or left out, it was a conscious design choice, re-confirmed a number of times. The choices might not accord to your (or my) preferences, they might change in the future based on evolving needs, but for now they are what they are, and they are there for a reason.

Of course adherence to standards is not mandatory, but I'd really suggest that standards are not pick-and-choose types of things: Either the output is supposed to be OSI, then it should try to conform, or it is not, then maybe it should not be named OSI output.

Anything else just leads to interoperability problems, of which we already have enough as it stands, given the lack of definition in standards. Deviating where standards are clear is not going to improve the picture.

**Comment by pmai** [comment]
> I will think about an idea to indicate static information to be present in the first frame only in mcap multi trace files and indicate that properly. I think it would boost rendering performance as well. I guess the first step is to indicate what information is allowed to be static in a simulation run. I will come back with an idea

We already have ways of separating out static information in traces via the GroundTruthInit message, as well as the StreamingUpdate messages for more interactive use cases. I do not foresee the need or the willingness to invent yet another way of separating out static information in yet another ad-hoc underspecified way. Especially various users at BMW were insistent that special-casing the first frame of simulation is not at all useful as it causes huge problems in real-time simulation use cases - beside the obvious definitional problems (what about things that are not static, but are only present in the first frame for some reason, and related corner cases).

**Comment by eknabevcc** [comment]
The concern we see, regarding GroundTruth, is to include the complete static road information, including every single road mark, in every frame. Especially for the trace file. If this is currently required by the standard, then we simply think this is a shortcoming that (easily) could be improved. 

For our understanding, what is the recommended way to communicate the complete groundtruth (including the complete road network), let's say in a trace file? Is there any documentation (e.g. a User Guide) outlining the various options based on use cases?

When it comes to API usage, esmini now supports grabbing the complete GroundTruth structure at every frame.

**Comment by thomassedlmayer** [comment]
> For our understanding, what is the recommended way to communicate the complete groundtruth (including the complete road network), let's say in a trace file? Is there any documentation (e.g. a User Guide) outlining the various options based on use cases?

As per OSI documentation a GroundTruth message should contain "all simulated objects in the global coordinate system at consecutive time instances" ([see here](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/V3.5.0/interface/architecture/ground_truth.html)). The definition is not being restrictive in the sense that it has to hold the "whole" world but when using it inside a SensorView message it becomes clear that it should hold the content that is necessary to represent a coherent snapshot of reality for the intended use (for a sensor model, traffic participant model, etc. ([see here](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/V3.7.0/gen/structosi3_1_1SensorView.html#afbb7fb861761c9d12b6a49ef1f0b15be))).

As I said before, other tools use chunking to limit the data per frame and only output a part of the road network around a specific ego vehicle. The pruned world representation is then put into a SensorView top level message (instead of a top level GroundTruth message) that is specifically created for that ego vehicle (and probably some use case). I don't really know a use case where it is necessary to have the whole generically simulated world in one GroundTruth trace. For this you could also save the road network with OpenDrive or any other map format. The more prevalent use cases that use OSI require limited (but coherent!) snapshots of reality and usually only require the part that e.g. a sensor would be able to perceive. That is why there is the SensorViewConfiguration message, that enables to find an agreement between SensorView producer and consumer on e.g. what field of view around the ego vehicle should be contained in the SensorView.

So in summary, my recommended way of creating OSI dumps is to - for each frame - prune the simulated world to some (configurable) extent, put that into a GroundTruth message that is contained in a SensorView message and indicate which vehicle is the ego vehicle.

Otherwise, a kind of workaround option could be to create two trace files: One holding all static information as GroundTruth (equivalent to a GroundTruthInit message) but no dynamic data; and another trace file that only contains the dynamic stuff. If this separation is clearly indicated, it would at least resolve the risk of misinterpretation which we currently have. But this case is not covered by the OSI documentation because I don't even know if someone wants this. At least OSI is not designed for this. It is possible to have OSI GroundTruth trace files on its own but as you noticed, it then also just doesn't make sense to redundantly put the whole road network into every frame. It should be customized towards a certain host vehicle, which also has to be indicated in the data. I think, this is also a missing field in the GroundTruth export from esmini if I remember correctly.

**Comment by pmai** [comment]
> The concern we see, regarding GroundTruth, is to include the complete static road information, including every single road mark, in every frame. Especially for the trace file. If this is currently required by the standard, then we simply think this is a shortcoming that (easily) could be improved.

It's the easily in this sentence that is the nub of the problem: If it was easy (meaning no downsides with serious impact), it would have been done. Hence the at least four different ways that this can be handled:

- For the usual set of restricted scenarios (i.e. size of relevant map is not excessively large), just storing full GroundTruth is usually not excessive, especially for trace files that can be written/read lzma compressed, which usually deals with this kind of replication nicely enough.
- When the goal is to provide traces centered around an ego vehicle (for sensor or non-sensor uses), then the provision of SensorView with localized filtering is a good fit.
- Optionally, separate GroundTruthInit message with SensorView that indicates this in its SensorViewConfiguration can be written. With current trace file format this means two trace files, with upcoming MCAP-based multi-trace file support this can be handled more transparently in one trace file.
- For more global uses, i.e. not necessarily centered around an ego, like visualization, StreamingUpdate can be used in conjunction with a GroundTruth Init message.

See also Thomas' suggestions.

> For our understanding, what is the recommended way to communicate the complete groundtruth (including the complete road network), let's say in a trace file? Is there any documentation (e.g. a User Guide) outlining the various options based on use cases?

There is currently no best practice guide that would elaborate the options beyond the definitions in the standard itself. We'll see if we can provide some more written guidance as part of the move to multi-trace files.

> When it comes to API usage, esmini now supports grabbing the complete GroundTruth structure at every frame.

In that case just an option to store them as such in trace files would not seem to be out of the ordinary. When using lzma compressed trace files, this should not be hugely costly in most localized use cases. For very large road networks, one of the other options might be preferable.

**Comment by eknabevcc** [comment]
I guess we'll have to get more involved in OSI in order to gain solid understanding of the reasons why non-changing data must be replicated.

At least we've now understood that there is a need also for an option to replicate static data in trace files. We can definitely add an esmini option to include static data in all trace file frames. 

But to avoid adding useless options, can any stakeholder just confirm that it would solve the problem at hand? Or is it something else that stops the show? 

Finally, a few words on esmini OSI handling and supported work flow:
esmini is only populating the global [GroundTruth](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/V3.6.0/gen/structosi3_1_1GroundTruth.html) message. The consumer will see the complete world. From this data sensor models, outside esmini, can be populated using OSI SensorView or other ways. 

Obviously esmini could also populate SensorView messages. But since there was no standardized way in OpenSCENARIO to configure sensors, we decided to focus on GroundTruth only. As usual, this is just how it works now, it might change in future. Either by us based on internal need, or by external contributions.

**Comment by eknabevcc** [comment]
> ... Especially various users at BMW were insistent that special-casing the first frame of simulation is not at all useful as it causes huge problems in real-time simulation use cases - beside the obvious definitional problems (what about things that are not static, but are only present in the first frame for some reason, and related corner cases).

A clarification for the record, to avoid misunderstandings:

Using esmini API the initial groundtruth can (should) be fetched already after initialization of the scenario, before first simulation step. Something like this:

```c++
SE_Init("myscenario.xosc"...);

// Now when scenario is loaded, and all entities initial state applied, we can grab complete groundtruth
SE_UpdateOSIGroundTruth();
gt = reinterpret_cast<const osi3::GroundTruth*>(SE_GetOSIGroundTruthRaw());

// process groundtruth, e.g. store road network if needed
for (int j = 0; j < gt->lane_boundary_size(); j++)
{
   for (int k = 0; k < gt->lane_boundary(j).boundary_line_size(); k++)
   {
      ...
   }
}

// step the simulation
while (SE_GetQuitFlag() == 0)
{
   SE_StepDT(0.01f);

   // grab updated groundtruth, by default only dynamic parts
   SE_UpdateOSIGroundTruth();
   gt = reinterpret_cast<const osi3::GroundTruth*>(SE_GetOSIGroundTruthRaw());

   // process it, e.g:
   for (int j = 0; j < gt->moving_object().size(); j++)
   {
       ...
   }
}
```

This way there will be two phases: 1. Initialization (which will take some time anyway to load scenario) 2. Running the simulation. By handling the static groundtruth already in the initialization phase we not only avoid irregular step durations but also reducing the duration of each step, in other words optimizing simulation performance. Whenever possible this is the recommended scheme. 

If the complete groundtruth is needed via API continuously, just change any:
```c++
SE_UpdateOSIGroundTruth();
```
to
```c++
SE_UpdateOSIGroundTruth(true);
```

**Comment by eknabevcc** [comment]
We see no issues with this scheme. Anyway, even if there are other approaches that could work, we believe that this is the straight forward way that users most likely will go for - hence we care for the default behavior (e.g. performance and trace file size). 

As stated previously, already we provide an API flag in case users need the complete content. 

The unanswered question is: Would an option for also put static data in each trace file entry would solve a problem? In which case we can add such flag.

**Comment by thomassedlmayer** [comment]
> We see no issues with this scheme. Anyway, even if there are other approaches that could work, we believe that this is the straight forward way that users most likely will go for - hence we care for the default behavior (e.g. performance and trace file size).

So in case you want to keep the current behavior for `--osi_file`, I think it would be great if the exact deviation from the ASAM standard was clearly stated. People should not be fooled into thinking that they are writing an OSI trace file that conforms to the standard.

The esmini user guide says that the "[...] format is according to the OSI standard [...]". But if this was true it would mean that from the second frame on there are no lanes present anymore. Isn't this a problem?

> The unanswered question is: Would an option for also put static data in each trace file entry would solve a problem? In which case we can add such flag.

Yes, it would actually enable esmini to write an interoperable OSI trace file without having to deal with assumptions on the static/dynamic issue we tried to get across.

Currently, I can only take your word (or dive into the code) to find out, what parts are considered static or dynamic in the first frame. I assume that it is just lanes and lane boundaries (?) but this is not indicated anywhere if I'm not mistaken?

**Comment by eknabevcc** [comment]
@pmai 
I now see that you refer to the trace file, not API usage which we addressed in the example code.

Maybe coming from different work flows makes the discussion a bit out of focus. So far we've seen trace file being a result/output from simulation, something that might be analyzed or reviewed in a post process. But from your input we sense that it's being used also in the simulation itself? Then I see how separate trace files could be useful - one for initialization (phase 1) and one for execution  (phase 2).

We'll discuss internally if this would be an useful option. 

Meanwhile, since we don't use and know the details of OSMP, it would be helpful with a brief description of how it works wrt OSI tracefiles. Does it READ two tracefiles, one with only the initial states, and a second containing all updates from the simulation? Does it also produce a new OSI file based on the first and some additional simulated vehicles?

Maybe an online meeting could be helpful to align understanding. We are a bit allergic to academic discussions, but highly motivated helping users getting the most out from the tools and ASAM standards.

**Comment by eknabevcc** [comment]
Static data is static, by nature it will not disappear. That's kind of the point. It only needs to be processed once.

In esmini the following OSI entities is treated as static:
- OpenDRIVE road objects
- OpenSCENARIO Misc objects (dont't move)
- Road lanes (center, boundaries, connectivity...)
- Geo reference

The categories of the first two are similar. Examples: obstacle, building, tree, parkingspace...

As a simple example, running cut-in.xosc at 10 ms timestep result:

- tracefile with static data once: 2.25 MB
- tracefile with static data always: 181.17 MB

The replication of static non-changing data adds 179 MB or 8000% or 80 times the size, without bringing any additional value. This can't make sense. 

Initially esmini did output static data always. Changing that behavior was a low hanging fruit in the quest for performance and scale. So basically we highly recommend it for any tool (register static data once). 

That said, thanks for input. We'll add an option to store static data always, since it will enable OSI files from esmini to work in other tools. If not already clear we'll clarify in the documentation both the default behavior and the option.

**Comment by thomassedlmayer** [comment]
Thank you for the clarifications! I'm looking forward to the new option! And I'm sorry for the extensive discussion but I think we had quite different perspectives on the topic.

> Maybe coming from different work flows makes the discussion a bit out of focus. So far we've seen trace file being a result/output from simulation, something that might be analyzed or reviewed in a post process. But from your input we sense that it's being used also in the simulation itself? Then I see how separate trace files could be useful - one for initialization (phase 1) and one for execution (phase 2).

I think it should be possible to treat trace files exactly like messages that are exchanged between simulation components, e.g. replay a simulation in another tool for purposes like visualization, validation, or whatever.

> Initially esmini did output static data always. Changing that behavior was a low hanging fruit in the quest for performance and scale. So basically we highly recommend it for any tool (register static data once).

By the way, I see no issues with esmini internally handling static content like this, especially since a large portion of OpenDrive and OpenScenario content is probably static by definition. My point all along was just about the OSI export because - by default - in OSI nothing is static if not indicated according to the mentioned valid options. Tools that interpret OSI can and should not simply assume that some things are static. That's why I said, lanes/lane boundaries disappear in the second frame. They obviously don't disappear in esmini but they would in a tool that's reading such an OSI trace.

Also here are some insights into the lzma compression (which is officially supported by the [osi3trace](https://github.com/OpenSimulationInterface/open-simulation-interface/tree/master/osi3trace) package btw): Compressing an OSI 130MB SensorData trace file (1500 frames with hundreds of radar detections and about 50 detected objects per frame) reduces the file size down to about 23MB. So it might be worth having a look at that for the new option?

> Meanwhile, since we don't use and know the details of OSMP, it would be helpful with a brief description of how it works wrt OSI tracefiles. Does it READ two tracefiles, one with only the initial states, and a second containing all updates from the simulation? Does it also produce a new OSI file based on the first and some additional simulated vehicles?
Maybe an online meeting could be helpful to align understanding. We are a bit allergic to academic discussions, but highly motivated helping users getting the most out from the tools and ASAM standards.

I'd leave this to @pmai as I'm sure he has more insight into OSMP.

**Comment by eknabevcc** [comment]
Extended options included in release [v2.46.0](https://github.com/esmini/esmini/releases/tag/v2.46.0)

It is now possible to always include static data also in trace file, not only via API.

See [v2.46.0 release notes](https://github.com/esmini/esmini/blob/master/release_notes.md) and [User guide - OSI data](https://esmini.github.io/#_osi_data) for details.
